pkg/codegen/testing/test/testdata/regress-py-14539/python-extras/tests/test_codegen.py

```diff
@@ -0,0 +1,133 @@
+
+
+class MyMocks(pulumi.runtime.Mocks):
+    def call(self, args):
+        return {}
+    def new_resource(self, args):
+        assert args.inputs == {
+            "bootDisk": {
+                "initializeParams": {
+                    "imageName": "debian-cloud/debian-11",
+                },
+            },
+        }
+        return "foo", args.inputs
+
+
+def identity(x):
+    return x
+
+
+def wrap_output(x):
+    return pulumi.Output.from_input(x)
+
+
+def wrap_future(x):
+    future = asyncio.Future()
+    future.set_result(x)
+    return future
+
+
+@pytest.mark.parametrize("boot_disk_wrap", [identity, wrap_output, wrap_future])
+@pytest.mark.parametrize("initialize_params_wrap", [identity, wrap_output, wrap_future])
+@pytest.mark.parametrize("image_name_wrap", [identity, wrap_output, wrap_future])
+@pytest.mark.parametrize("initialize_params", ["initialize_params", "initializeParams"])
+@pytest.mark.parametrize("image_name", ["image_name", "imageName"])
+@pulumi.runtime.test
+def test_dict(
+    my_mocks,
+    boot_disk_wrap,
+    initialize_params_wrap,
+    image_name_wrap,
+    initialize_params,
+    image_name,
+):
+    Instance(
+        "instance-1",
+        boot_disk=boot_disk_wrap({
+            initialize_params: initialize_params_wrap({
+                image_name: image_name_wrap("debian-cloud/debian-11"),
+            }),
+        }),
+    )
+
+
+@pytest.mark.parametrize("boot_disk_wrap", [identity, wrap_output, wrap_future])
+@pytest.mark.parametrize("initialize_params_wrap", [identity, wrap_output, wrap_future])
+@pytest.mark.parametrize("image_name_wrap", [identity, wrap_output, wrap_future])
+@pulumi.runtime.test
+def test_input_types(
+    my_mocks,
+    boot_disk_wrap,
+    initialize_params_wrap,
+    image_name_wrap,
+):
+    Instance(
+        "instance-1",
+        boot_disk=boot_disk_wrap(InstanceBootDiskArgs(
+            initialize_params=initialize_params_wrap(InstanceBootDiskInitializeParamsArgs(
+                image_name=image_name_wrap("debian-cloud/debian-11"),
+            )),
+        )),
+    )
+
+    Instance(
+        "instance-2",
+        InstanceArgs(
+            boot_disk=boot_disk_wrap(InstanceBootDiskArgs(
+                initialize_params=initialize_params_wrap(InstanceBootDiskInitializeParamsArgs(
+                    image_name=image_name_wrap("debian-cloud/debian-11"),
+                )),
+            )),
+        ),
+    )
+
+@pulumi.runtime.test
+def test_mutate_input_types(my_mocks):
+    initialize_params_args = InstanceBootDiskInitializeParamsArgs()
+    initialize_params_args.image_name = "debian-cloud/debian-11"
+
+    boot_disk_args = InstanceBootDiskArgs(initialize_params="bad-initialize-params")
+    boot_disk_args.initialize_params = initialize_params_args
+    Instance("instance-1", boot_disk=boot_disk_args)
+
+    instance_args = InstanceArgs(boot_disk="bad-boot-disk")
+    instance_args.boot_disk = boot_disk_args
+
+    Instance(
```

Comment:
Odd indent here, also does bad-boot-disk above trigger the assert to cause an error here?

Reply:
Great question! It does, but I was experiencing some oddness where `with pytest.raises(...):` would refuse to catch the AssertionError triggered in the Mock and after trial and searching for a bit, I didn't find any good work arounds and it didn't seem to make sense to block on it.

Reply:
It triggers an assert, but the assert doesn't bubble up from the call-site so writing a test to catch an assert thrown in the Mock is what I was having difficulty with.

Code:
```python
    instance_args.boot_disk = "bad-boot-disk"
    with pytest.raises(Exception):
        Instance(
            "instance-2",
            instance_args,
        )
```
Test results:
```
            io.go:432: ___________________________ test_mutate_input_types ____________________________
            io.go:432: 
            io.go:432: my_mocks = <test_codegen.MyMocks object at 0x7f94b31f7610>
            io.go:432: 
            io.go:432:     @pulumi.runtime.test
            io.go:432:     def test_mutate_input_types(my_mocks):
            io.go:432:         initialize_params_args = InstanceBootDiskInitializeParamsArgs()
            io.go:432:         initialize_params_args.image_name = "debian-cloud/debian-11"
            io.go:432:     
            io.go:432:         boot_disk_args = InstanceBootDiskArgs(initialize_params="bad-initialize-params")
            io.go:432:         boot_disk_args.initialize_params = initialize_params_args
            io.go:432:         Instance("instance-1", boot_disk=boot_disk_args)
            io.go:432:     
            io.go:432:         instance_args = InstanceArgs(boot_disk="bad-boot-disk")
            io.go:432:         instance_args.boot_disk = boot_disk_args
            io.go:432:     
            io.go:432:         Instance(
            io.go:432:             "instance-2",
            io.go:432:             instance_args,
            io.go:432:         )
            io.go:432:     
            io.go:432:         instance_args.boot_disk = "bad-boot-disk"
            io.go:432:     
            io.go:432: >       with pytest.raises(Exception):
            io.go:432: E       Failed: DID NOT RAISE <class 'Exception'>
            io.go:432: 
            io.go:432: tests/test_codegen.py:139: Failed
            io.go:432: 
            io.go:432: During handling of the above exception, another exception occurred:
            io.go:432: 
            io.go:432: self = <test_codegen.MyMocks object at 0x7f94b31f7610>
            io.go:432: args = <pulumi.runtime.mocks.MockResourceArgs object at 0x7f94b208f650>
            io.go:432: 
            io.go:432:     def new_resource(self, args):
            io.go:432: >       assert args.inputs == {
            io.go:432:             "bootDisk": {
            io.go:432:                 "initializeParams": {
            io.go:432:                     "imageName": "debian-cloud/debian-11",
            io.go:432:                 },
            io.go:432:             },
            io.go:432:         }
            io.go:432: E       AssertionError: assert {'bootDisk': 'bad-boot-disk'} == {'bootDisk': .../debian-11'}}}
            io.go:432: E         Differing items:
            io.go:432: E         {'bootDisk': 'bad-boot-disk'} != {'bootDisk': {'initializeParams': {'imageName': 'debian-cloud/debian-11'}}}
            io.go:432: E         Use -v to get more diff
            io.go:432: 
            io.go:432: tests/test_codegen.py:42: AssertionError
```